[shopsys] add frontend-api bundle with category tree

grossmannmartin opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| This PR introduces GraphQL Frontend API and use it for obtain category tree.
|New feature| Yes <!-- Do not forget to update docs/ -->
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| ... <!-- Write "closes #123" for the issue to be closed automatically during merge -->
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

grossmannmartin commented on packages/frontend-api/README.md:
```diff
@@ -0,0 +1,55 @@
+
+This bundle for [Shopsys Framework](https://www.shopsys.com) adds frontend API using [overblog/GraphQLBundle](https://github.com/overblog/GraphQLBundle).
+The bundle is dedicated for projects based on Shopsys Framework (i.e. created from [`shopsys/project-base`](https://github.com/shopsys/project-base)) exclusively.
+This repository is maintained by [shopsys/shopsys] monorepo, information about changes is in [monorepo CHANGELOG.md](https://github.com/shopsys/shopsys/blob/master/CHANGELOG.md).
```
Well, this is quite new, but I believe it should be truly master according to PR #1436. 
When version 9.0 will be released all the changes should be in the master version of the changelog file and having it differently could lead to misleading information due to viewing the not updated version.

grossmannmartin commented on packages/frontend-api/src/Resources/config/services.yml:
```diff
@@ -0,0 +1,16 @@
+    Shopsys\FrontendApiBundle\:
+        resource: '../../**/*{Resolver}.php'
+
+    Shopsys\FrontendApiBundle\Routing\EnabledOnCurrentDomainChecker: ~
```
Good catch, thanks!

grossmannmartin replied:
Well, `%shopsys_frontend_api.enabled_domains%` is not registered as parameter. 
Into this service it's passed in `ShopsysFrontendApiExtension.php`

TomasLudvik replied:
Okay that is something I do not like, because when you are looking for arguments, that are wired to the service, you look into services.yml and you find nothing, then you need to start thinking, where it might be.

grossmannmartin replied:
I rewrote it to the standard parameter in a14a795771f082c8dbbc0d18cf4d3cc0d33eee39 and 19baa4e863e844360de88dd38dffe3d6e0d301fa
(documentation and upgrade notes are changed too in 62c5a3973b1c1d78372ed2a01da6336b1580bffe and ee698a5e1521a7dea930e907e14869dd9fa67a56)

grossmannmartin commented on upgrade/UPGRADE-v9.0.0-dev.md:
```diff
@@ -33,6 +33,37 @@ There you can find links to upgrade notes for other versions too.
         - `src/Resources/views/Admin/Content/Category/detail.html.twig`
         - `src/Resources/views/Admin/Content/Product/detail.html.twig`
 - add [`app/getEnvironment.php`](https://github.com/shopsys/shopsys/blob/9.0/project-base/app/getEnvironment.php) file to your project ([#1368](https://github.com/shopsys/shopsys/pull/1368))
+- introduce [Frontend API](https://github.com/shopsys/shopsys/blob/9.0/docs/frontend-api/introduction-to-frontend-api.md) to your project ([#](https://github.com/shopsys/shopsys/pull/)):
```
Applied manually in 549f3a310

grossmannmartin commented on upgrade/UPGRADE-v9.0.0-dev.md:
```diff
@@ -33,6 +33,37 @@ There you can find links to upgrade notes for other versions too.
         - `src/Resources/views/Admin/Content/Category/detail.html.twig`
         - `src/Resources/views/Admin/Content/Product/detail.html.twig`
 - add [`app/getEnvironment.php`](https://github.com/shopsys/shopsys/blob/9.0/project-base/app/getEnvironment.php) file to your project ([#1368](https://github.com/shopsys/shopsys/pull/1368))
+- introduce [Frontend API](https://github.com/shopsys/shopsys/blob/9.0/docs/frontend-api/introduction-to-frontend-api.md) to your project ([#](https://github.com/shopsys/shopsys/pull/)):
```
👍 applied manually in 549f3a310

grossmannmartin commented on packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * I case no domain is selected, check if enabled on Main Admin Domain
```
(hopefully) improved in 156b538c1

TomasLudvik commented on docs/frontend-api/introduction-to-frontend-api.md:
```diff
@@ -1,21 +1,148 @@
+In case you want to add the package into an already existing project, you should follow [upgrade instructions](https://github.com/shopsys/shopsys/blob/9.0/upgrade/UPGRADE-v9.0.0-dev.md)
 
 ## Configuration
+<!--- TODO add after configuration refactoring -->
```
TODO here

grossmannmartin replied:
Content added in ee698a5e1521a7dea930e907e14869dd9fa67a56

TomasLudvik commented on docs/frontend-api/introduction-to-frontend-api.md:
```diff
@@ -0,0 +1,148 @@
+Frontend API package also integrates graphical interactive in-browser GraphQL IDE [GraphiQL](https://github.com/graphql/graphiql/tree/master/packages/graphiql#readme).
+With it, you can debug your query easily and also browse the endpoint documentation for available objects, fields and their meaning.
+
+You can access it in development mode on `http://127.0.0.1/graphql/graphiql` (respectively `http://127.0.0.1/graphql/graphiql` for the second domain).
```
```suggestion
You can access it in development mode on `http://127.0.0.1/graphql/graphiql` (respectively `http://127.0.0.2/graphql/graphiql` for the second domain).
```

TomasLudvik commented on docs/frontend-api/introduction-to-frontend-api.md:
```diff
@@ -0,0 +1,148 @@
+    inherits:
+        - 'CategoryDecorator'   # Inherits from the decorator defined earlier
+    config:
+        fields:           # Fields from the decorator are added automatically
```
These comments are weirdly aligned

grossmannmartin replied:
fixed in 34d0754fb

TomasLudvik commented on packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * If no domain is selected method returns if it's enabled for the first domain instead.
```
I think it would be better to add this info somehow in methods name, because right now this behavior is not clear and you need to read that comment which you might not ...

grossmannmartin replied:
Changed to `isEnabledOnDomain(int $domainId)` in 4d0fda5e0b9ee10fa5af21bba41a3e2763cdf86b

grossmannmartin commented on docs/frontend-api/introduction-to-frontend-api.md:
```diff
@@ -0,0 +1,21 @@
+# Introduction to Frontend API
+
+Shopsys Framework Frontend API is an interface to the application that is dedicated for integration with your third party systems systems (like IS, storage system, ERP, ...).
```
resolved manually in 7fe0f9ae0

TomasLudvik commented on packages/frontend-api/src/Routing/FrontendApiRouteLoader.php:
```diff
@@ -39,7 +48,11 @@ public function loadRoutes(): RouteCollection
+            if (!$this->enabledOnDomainChecker->isEnabledOnDomain($this->domain->getId())) {
+                return $routes;
+            }
+        } catch (NoDomainSelectedException $e) {
```
Where is this exception thrown?

grossmannmartin replied:
`$this->domain->getId()` throws this exception :)

TomasLudvik commented on docs/frontend-api/introduction-to-frontend-api.md:
```diff
@@ -0,0 +1,157 @@
+# Introduction to Frontend API
+
+Shopsys Framework Frontend API is an interface to the application that is used for integration with external shop frontend, for example, JS storefront od mobile app.
```
```suggestion
Shopsys Framework Frontend API is an interface to the application that is used for integration with external store frontend, for example, JS storefront od mobile app.
```

grossmannmartin commented:
> Hello, from testing:
> 
> * 127.0.0.1:8000/graphql does not work... It would be nice if I would not need exact address
> * if I have not enabled API on domain I will get this error: "Symfony\Component\HttpKernel\Exception\NotFoundHttpException: No route found for "GET /graphql/"" something with better description would be nice
> * all errors are returned as HTML that will be hard to parse in javascript - JSON would be highly appreciated

Thanks!

- GET request is by default redirected, but POST (request made from Postman, for example) can't be redirected and therefore the trailing slash is needed. The only way I can think of is to not specify route prefix in project-base and hardcode both variants in `FrontendApiRouteLoader`. But I'm not sure if it's worth it (especially when different behavior of routes with/without slash seems like standard behavior – http://127.0.0.1:8000/registration/ vs. http://127.0.0.1:8000/registration)
- Yes, it's standard 404 page because there is no such route. I believe it's proper behavior, because no such thing actually exist on the domain in that scenario.
- I added `ErrorHandlerListener` in ad7f2d449 to show error messages from overblog_graphql_endpoint nicely.

grossmannmartin commented:
Hi @TomasLudvik,
I found a bug in enabling routes logic (caching routes could be a problem if one domain was disabled), so I changed the way how the endpoint is enabled/disabled. It's now checked at the controller level and with it, I was able to solve your comment

> if I have not enabled API on domain I will get this error: "Symfony\Component\HttpKernel\Exception\NotFoundHttpException: No route found for "GET /graphql/"" something with better description would be nice

work is done mainly in 59a7ba3 (but the last three commits are related).
Can you please check it out?

grossmannmartin commented:
> * I think it is not necessary to secure access to GraphiQL in dev environment, because if you try to use it it will tell you that GraphQL is not enabled on domain. In that case I would remove services_dev.yml and other connected changes.

Thanks, you're right, it's better without the services_dev :)
Fixed in 43e1420f26ab88fa78acd148661d0e7c19ef9f59. And I also moved `EnabledOnDomainChecker` to namespace `Component` instead of `Routing`, if you agree.
